dask_planner/src/sql/optimizer/join_reorder.rs

```diff
@@ -0,0 +1,385 @@
+//! Join reordering based on the paper "Improving Join Reordering for Large Scale Distributed Computing"
+//! https://ieeexplore.ieee.org/document/9378281
+
+use std::collections::HashSet;
+
+use datafusion_common::{Column, Result};
+use datafusion_expr::{Expr, Join, JoinType, LogicalPlan, LogicalPlanBuilder};
+use datafusion_optimizer::{utils, utils::split_conjunction, OptimizerConfig, OptimizerRule};
+
+use crate::sql::table::DaskTableSource;
+
+pub struct JoinReorder {
+    /// Maximum number of fact tables to allow in a join
+    max_fact_tables: usize,
+    /// Ratio of the size of the dimension tables to fact tables
+    fact_dimension_ratio: f64,
+    /// Whether to preserve user-defined order of unfiltered dimensions
+    preserve_user_order: bool,
+    /// Constant to use when determining the number of rows produced by a
+    /// filtered relation
+    filter_selectivity: f64,
+}
+
+impl Default for JoinReorder {
+    fn default() -> Self {
+        Self {
+            max_fact_tables: 2,
+            // FIXME: fact_dimension_ratio should be 0.3
+            fact_dimension_ratio: 0.7,
```

Comment:
Ideally, this should be 0.3, but queries 17, 25, 29, and 85 currently fail without a stricter ratio. This suggests that more work needs to be done with reordering fact-to-fact joins.